pkg/cluster/logs/logs.go

```diff
@@ -152,3 +145,92 @@ func coalesce(fns ...errFn) error {
 	}
 	return nil
 }
+
+// untar reads the tar file from r and writes it into dir.
+// adapted from https://github.com/golang/build/blob/master/cmd/buildlet/buildlet.go
+func untar(r io.Reader, dir string) (err error) {
+	nFiles := 0
+	madeDir := map[string]bool{}
+	defer func() {
+		if err == nil {
+			log.Printf("extracted tarball into %s: %d files, %d dirs", dir, nFiles, len(madeDir))
+		} else {
+			log.Printf("error extracting tarball into %s after %d files, %d dirs: %v", dir, nFiles, len(madeDir), err)
+		}
+	}()
+	tr := tar.NewReader(r)
+	for {
+		f, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return errors.Wrapf(err, "tar reading error: %v", err)
+		}
+		if f.Typeflag == tar.TypeXGlobalHeader {
+			// golang.org/issue/22748: git archive exports
+			// a global header ('g') which after Go 1.9
+			// (for a bit?) contained an empty filename.
+			// Ignore it.
+			continue
+		}
+		if !validRelPath(f.Name) {
+			return errors.Wrapf(err, "tar file contained invalid name %q", f.Name)
+		}
+		rel := filepath.FromSlash(f.Name)
+		abs := filepath.Join(dir, rel)
+
+		fi := f.FileInfo()
+		mode := fi.Mode()
+		switch {
+		// we want to extract all files and symlinks
+		case mode.IsRegular():
+			// Make the directory. This is redundant because it should
+			// already be made by a directory entry in the tar
+			// beforehand. Thus, don't check for errors; the next
+			// write will fail with the same error.
+			dir := filepath.Dir(abs)
+			if !madeDir[dir] {
+				if err := os.MkdirAll(filepath.Dir(abs), 0755); err != nil {
+					return err
+				}
+				madeDir[dir] = true
+			}
+			wf, err := os.OpenFile(abs, os.O_RDWR|os.O_CREATE|os.O_TRUNC, mode.Perm())
+			if err != nil {
+				return err
+			}
+			n, err := io.Copy(wf, tr)
+			if closeErr := wf.Close(); closeErr != nil && err == nil {
+				err = closeErr
+			}
+			if err != nil {
+				return errors.Errorf("error writing to %s: %v", abs, err)
+			}
+			if n != f.Size {
+				return errors.Errorf("only wrote %d bytes to %s; expected %d", n, abs, f.Size)
+			}
+			nFiles++
+		case mode.IsDir():
+			if err := os.MkdirAll(abs, 0755); err != nil {
+				return err
+			}
+			madeDir[abs] = true
+		case mode&os.ModeSymlink != 0:
+			// TODO: ignore these for now. They were breaking x/build tests.
+			// Implement these if/when we ever have a test that needs them.
```

Comment:
I'm not sure we can just "adapt" this much code under our own license...

Reply:
this can be used instead:
https://medium.com/@skdomino/taring-untaring-files-in-go-6b07cf56bc07

Reply:
Yeah, indeed seems too restrictive https://github.com/golang/build/blob/master/LICENSE, however is not a big chunk of code and there are other several projects using the same structure. Will refactor it, I don't think it can be considered as derivatives work

Reply:
updated de untar code based on @neolit123 suggestion